src/meta/service/src/ddl_service.rs

```diff
@@ -1364,6 +1368,178 @@ impl DdlService for DdlServiceImpl {
+
+        assert_eq!(fragment_graph.dependent_table_ids.len(), 1);
+        assert!(
+            risingwave_common::catalog::TableId::from(fragment_graph.dependent_table_ids[0])
+                .is_placeholder()
+        );
+        fragment_graph.dependent_table_ids[0] = table_catalog.id;
+        for fragment in fragment_graph.fragments.values_mut() {
+            stream_graph_visitor::visit_fragment_mut(fragment, |node| match node {
+                NodeBody::StreamScan(scan) => {
+                    scan.table_id = table_catalog.id;
+                    if let Some(table_desc) = &mut scan.table_desc {
+                        assert!(
+                            risingwave_common::catalog::TableId::from(table_desc.table_id)
+                                .is_placeholder()
+                        );
+                        table_desc.table_id = table_catalog.id;
+                        table_desc.maybe_vnode_count = table_catalog.maybe_vnode_count;
+                    }
+                    if let Some(table) = &mut scan.arrangement_table {
+                        assert!(
+                            risingwave_common::catalog::TableId::from(table.id).is_placeholder()
+                        );
+                        *table = table_catalog.clone();
+                    }
+                }
+                NodeBody::BatchPlan(plan) => {
+                    if let Some(table_desc) = &mut plan.table_desc {
+                        assert!(
+                            risingwave_common::catalog::TableId::from(table_desc.table_id)
+                                .is_placeholder()
+                        );
+                        table_desc.table_id = table_catalog.id;
+                        table_desc.maybe_vnode_count = table_catalog.maybe_vnode_count;
+                    }
+                }
+                _ => {}
+            });
+        }
+
+        let table_id = table_catalog.id as ObjectId;
+        let dependencies = HashSet::from_iter([table_id]);
+        let stream_job = StreamingJob::Sink(sink);
+        let res = self
+            .ddl_controller
+            .run_command(DdlCommand::CreateStreamingJob {
+                stream_job,
+                fragment_graph,
+                dependencies,
+                specific_resource_group: None,
+                if_not_exists,
+            })
+            .await;
+
+        if res.is_err() {
+            let _ = self
+                .ddl_controller
+                .run_command(DdlCommand::DropStreamingJob {
+                    job_id: StreamingJobId::Table(None, table_id as _),
+                    drop_mode: DropMode::Cascade,
+                })
+                .await
+                .inspect_err(|err| {
+                    tracing::error!(error = %err.as_report(),
+                        "Failed to clean up table after iceberg sink creation failure",
+                    );
+                });
```

Comment:
If the iceberg sink is failed to create, I think we need to drop the iceberg table as well, because in the frontend, we have created the iceberg table.

Reply:
Yes, actually we need to drop the iceberg table at any point the rpc call fails. Let me move this part to the frontend, Although there is still no transaction guarantee, I will try to avoid the logic of pre-create iceberg catalog and migrate it to meta.